tests/integrationv2/test_buffered_send.py

```diff
@@ -0,0 +1,161 @@
+
+    starting_client_send_marker = other_provider.get_send_marker()
+
+    starting_server_send_marker = "YTREWQ"
+    client_inital_app_data = to_bytes(starting_server_send_marker)
+    client_close_marker = server_sent_final = "QWERTY"
+    data_bytes_server = data_bytes(SEND_DATA_SIZE) + to_bytes(server_sent_final)
+
+    port = next(available_ports)
+
+    client_options = ProviderOptions(
+        mode=Provider.ClientMode,
+        port=port,
+        data_to_send=client_inital_app_data,
+        insecure=True,
+        protocol=protocol)
+
+    extra_flags = ['--buffered-send', buffer_size]
+    if fragment_preference is not None:
+            extra_flags.append(fragment_preference)
+
+    server_options = ProviderOptions(
+        mode=Provider.ServerMode,
+        port=port,
+        cipher=cipher,
+        data_to_send=data_bytes_server,
+        insecure=True,
+        protocol=protocol,
+        key=certificate.key,
+        cert=certificate.cert,
+        extra_flags=extra_flags)
+
+    server = managed_process(provider, server_options, send_marker=[starting_server_send_marker])
+    client = managed_process(other_provider, client_options,
+        send_marker=[starting_client_send_marker], close_marker=client_close_marker)
+
+    for results in client.get_results():
+        # for small buffer sizes the received data will not be contiguous on stdout
+        assert(is_subsequence(
+            data_bytes_server.decode(encoding="ascii", errors="backslashreplace"),
+            results.stdout.decode(encoding="ascii", errors="backslashreplace")))
+        results.assert_success()
+
+    for results in server.get_results():
+        # the server should close without error
+        # but there is otherwise nothing of interest on stdout 
+        results.assert_success()
+
+
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("cipher", ALL_TEST_CIPHERS, ids=get_parameter_name)
+@pytest.mark.parametrize("other_provider", [S2N], ids=get_parameter_name)
+@pytest.mark.parametrize("provider", [S2N, OpenSSL], ids=get_parameter_name)
+@pytest.mark.parametrize("protocol", PROTOCOLS, ids=get_parameter_name)
+@pytest.mark.parametrize("certificate", TEST_CERTS, ids=get_parameter_name)
+@pytest.mark.parametrize("buffer_size", SEND_BUFFER_SIZES, ids=get_parameter_name)
+@pytest.mark.parametrize("fragment_preference", FRAGMENT_PREFERENCE, ids=get_parameter_name)
+def test_s2n_buffered_send_client(managed_process, cipher, other_provider, provider, protocol, certificate, buffer_size, fragment_preference):
+    # Communication Timeline
+    # Client [S2N]                 | Server [S2N|OpenSSL]
+    # Handshake                    | Handshake
+    #  Handshake finish indicated
+    # Send Data Bytes to Server    | Receive the Data Bytes
+    #  stresses the send buffer
+    #  center of test
+    # Send Server Close Marker     | Receive the close Marker
+    # Close                        | Close
```

Comment:
Can the server use some part of the test data as the close marker?

Reply:
It could indeed. It is important that it use a small enough part of it that it is printed contiguously on stdout.

In the process of working on this PR I was frequently changing the size of the data sent to diagnose an out of memory error -- in turn effecting what the close marker would be.

I could programmatically slice off the the last few bytes of the data and use that as the close marker -- I think it would work without issue.

Reply:
After experimenting with this for a while I've learned a few things:

First `data_bytes`, that we use to generate application data, cycles through the non-printable ascii characters repeatedly. This means that we can't use less than all the bytes (minus the number of non-printable ascii characters) for the close marker. Any reasonably sized suffix slice of the application data appears in the data not as a suffix; making it useless as a close marker.

So we have to use all the send data or insert a suffix not as a suffix of `data_bytes`.

Using all the send data mysteriously fails (even with higher debug levels): https://github.com/harrisonkaiser/s2n-tls/blob/3b6fa226141dd3e343342b9e3eca3e2d1fd582e6/tests/integrationv2/test_buffered_send.py#L90

```
[gw2] [  2%] PASSED test_buffered_send.py::test_s2n_server_buffered_send[None-1031-RSA_2048_SHA512-TLS1.0-OpenSSL-AES256-SHA]
test_buffered_send.py::test_s2n_server_buffered_send[None-1031-RSA_2048_SHA512-TLS1.0-S2N-ECDHE-RSA-AES256-SHA]
test_buffered_send.py::test_s2n_server_buffered_send[None-1031-RSA_4096_SHA256-TLS1.2-GnuTLS-ECDHE-RSA-AES128-SHA]
test_buffered_send.py::test_s2n_server_buffered_send[None-1031-RSA_4096_SHA256-TLS1.2-GnuTLS-ECDHE-RSA-AES128-SHA256] ERROR: InvocationError for command /s2n-tls/tests/integrationv2/.tox/py39/bin/pytest -n 8 --maxfail=1 --reruns=0 --cache-clear -rpfsq --provider-version=openssl-3.0 --fips-mode=0 --no-pq=0 --log-cli-level=INFO test_buffered_send.py (exited with code -9 (SIGKILL)) (exited with code -9)
______________________________________________________________________________________________________________________ summary ______________________________________________________________________________________________________________________
ERROR:   py39: commands failed
```

Instead we should have a close marker that isn't generated by `data_bytes` and doesn't appear in any output of any provider. I chose a well commented random string.